A py script to combine multiple outputs of Evidence QC in one table

shadizaheri opened the pull request:
This is the python script created to add some of the metrics resulting from the EvidenceQC step to the out put data table.

VJalili commented:
Thank you, @shadizaheri!!! It reads well, I made a few changes, mostly styling and abstracting methods, but please feel free to revert any if you prefer otherwise. Specifically, I made the following changes.

- There were multiple methods for reading the outputs of different methods (one for each of melt, manta, and wham). These methods were mostly implementing a similar functionality except for one array index. I refactored this into one method and made the array index as an input argument that can be adjusted w.r.t. to the method.

- You are composing some column names using filenames; e.g., from `bwa.Manta.QC.outlier.high` (filename), you compose `Manta_high` (column name). I think setting column names in the code w.r.t to the methods you call is easier and less error-prone than relying on the template of the filename. So, I made some changes similar to the following. 

```python
# Example 1
# from:
outlier_manta = filename_manta.split(".")[-4] + "_" + filename_manta.split(".")[-1]

# to
col_name = get_col_name("manta", outlier_type)

def get_col_name(caller: str, outlier_type: str) -> str:
    return f"{caller}_{outlier_type}_outlier"

# Example 2
# from
read_wham_outlier(filename_high_wham)
def read_wham_outlier(filename):
    outlier_wham = filename.split(".")[-4] + "_" + filename.split(".")[-1]
    ...
    outlier_wham_df.columns = ["#ID", str(outlier_wham) + "_outlier"]

# to
read_outlier(filename_high_wham, get_col_name("wham", "high"))
def read_outlier(filename, col_name):
    ...
    outlier_df.columns = ["#ID", col_name]
```

- I defined constant variable with its value set to `"#ID"`, and replaced all the references of `"#ID"` with the const var. This makes it less error-prone. 

- I fixed some issues with the command line args (optional vs. required, and the `-` to `_` conversion), and added a simple help message. Please feel free to extend the help messages if needed.

VJalili commented on src/sv-pipeline/scripts/make_evidence_qc_table.py:
```diff
@@ -0,0 +1,264 @@
+        df[ID_COL] = df.apply(lambda _: "NA", axis=0)
+        outlier_sample = df.pivot_table(columns=[ID_COL], aggfunc="size").astype(int)
+    else:
+        df[ID_COL] = df["Outlier_Sample"].str.split("/", expand=True)[idx].str.split(".", expand=True)[0]
```
I am not sure how extracting sample ID from its URI is portable/reliable. But I'll leave it to you to decide how best the sample ID can be extracted.

shadizaheri replied:
@VJalili using `os.path.split()` could also be an option, but it was not working here.

VJalili replied:
@shadizaheri do you have sample id in a different column of your data table? if so, that could be a better alternative, if not, it should be find to leave it as is.

VJalili replied:
That makes sense, thanks!!

VJalili commented on src/sv-pipeline/scripts/make_evidence_qc_table.py:
```diff
@@ -0,0 +1,264 @@
+    """
+    df = pd.read_csv(filename, sep="\t")
+    if df["Outlier_Sample"].empty:
+        df[ID_COL] = df.apply(lambda _: "NA", axis=0)
```
You had set the ID to empty string, I have seen issues with that in particular when a line is trimmed before split by a delimiter. So, I changed the empty string with `NA`; however, I'm not sure what is the best approach here. Do you need this row?

shadizaheri replied:
@VJalili I think we need this row so that it can return `'0'` for the empty files (the files that we did not have any types of outliers).

VJalili replied:
Thanks! That makes sense to me.

shadizaheri replied:
@epiercehoffman It is therefore necessary to check if the file is blank. Empty files should also include outlier columns. Even empty files need IDs when merging based on the `'#ID'` key at the end.  I think it will be more clear once you run the code using the empty and non-empty files.

epiercehoffman replied:
Do you end up with an extra row for `NA` in the case of an empty file due to the outer join later? I see an extra row with no ID in your example output, which we would want to remove

VJalili commented on src/sv-pipeline/scripts/make_evidence_qc_table.py:
```diff
@@ -0,0 +1,264 @@
+    ).fillna(0)  # "none"
+
+    # save the file
+    return output_df.to_csv(f"{output_prefix}evidence_qc_table.tsv", sep="\t", header=True, index=False)
```
what is returned here? I guess it is `None`, if so, you may remove `return`.

shadizaheri replied:
@VJalili This is supposed to return the final merged DataFrames in a "tsv" format.

VJalili replied:
Do you mean the filename?! I think since you're not using the returned value, it should be fine to remove `return`, but it should also be fine to leave it as is, it is up to you.

epiercehoffman replied:
Agreed, no need for a `return` since it is unused. Make sure to update the WDL with the expected output file name so that they match

shadizaheri replied:
Thank you @VJalili and @epiercehoffman ! This is a valid point. The only reason that I had the `'return'` here, was to test the script by outputting a tsv file. Once we test the script and have the wdl written, we can remove the `'return'` line from here.

epiercehoffman replied:
The return has no impact on whether the TSV file is written. You can delete the `return` keyword and keep the rest of the line and it will write the TSV file - for testing & for use in the workflow.

shadizaheri replied:
aha, now I see what you mean. Thanks for the comment and explanation. I removed `return`

shadizaheri reviewed (commented):
Thank you so much, @VJalili  for your comments and suggestions for the script's style. Very helpful.

VJalili reviewed (approved):
Thank you, @shadizaheri!!! The script reads very well. I tested it and it runs as we discussed; though I would leave it to @mwalker174 or @epiercehoffman to check if it implements the logic we discussed.

epiercehoffman commented on src/sv-pipeline/scripts/make_evidence_qc_table.py:
```diff
@@ -0,0 +1,264 @@
+    merged_dicts = Counter()
+    for counted in outlier_dicts:
+        merged_dicts.update(counted)
+    all_outliers = dict(merged_dicts)
```
When I did this I called `Counter()` on the concatenated sample ID columns from all the high outliers files, which seemed a little simpler/marginally more efficient because there were no dictionaries and you could count all of the callers at once

epiercehoffman commented on src/sv-pipeline/scripts/make_evidence_qc_table.py:
```diff
@@ -0,0 +1,264 @@
+    ).fillna(0)  # "none"
+
+    # save the file
+    return output_df.to_csv(f"{output_prefix}evidence_qc_table.tsv", sep="\t", header=True, index=False)
```
Let's just make the prefix required so we can add a delimiter here
```suggestion
    return output_df.to_csv(f"{output_prefix}.evidence_qc_table.tsv", sep="\t", header=True, index=False)
```

epiercehoffman commented on src/sv-pipeline/scripts/make_evidence_qc_table.py:
```diff
@@ -0,0 +1,264 @@
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "ploidy_test_filename",
+        help="Sets the filename containing ploidy tests.")
```
```suggestion
        "estimated_copy_number_filename",
        help="Sets the filename containing estimated copy numbers per contig.")
```

epiercehoffman commented on src/sv-pipeline/scripts/make_evidence_qc_table.py:
```diff
@@ -0,0 +1,264 @@
+        help="Sets the filename containing ploidy tests.")
+
+    parser.add_argument(
+        "median_cov_transposed_filename",
```
```suggestion
        "median_cov_filename",
```

epiercehoffman commented on src/sv-pipeline/scripts/make_evidence_qc_table.py:
```diff
@@ -0,0 +1,264 @@
+
+    parser.add_argument(
+        "--output-prefix",
+        default="",
```
```suggestion
```

shadizaheri replied:
@epiercehoffman I'm not sure if I understand the suggestion here. Can you please explain the suggestion, here?

epiercehoffman replied:
I think the prefix should be a required argument, so it does not need a default value

epiercehoffman reviewed (commented):
Great, this is looking very close to ready! I think it will be really helpful. Thanks @shadizaheri for your work on this. I think overall the code looks organized, readable, well-documented, and 99% correct (though I haven't had a chance to do a test run).

There's only one real change to the actual output values that I've requested, which is related to the rounded chrX copy number used for batching. My other suggestions are for code style / simplicity / efficiency.

I'd also like to see an example of the output. Can you share an example output table and the input files used for the test? I'd also like to test the script as part of the EvidenceQc workflow because there is some preparation of the inputs involved - let me know what your plan is for finalizing the WDL and integrating the changes across branches.

shadizaheri commented on src/sv-pipeline/scripts/make_evidence_qc_table.py:
```diff
@@ -0,0 +1,264 @@
+    """
+    df_non_diploid = pd.read_csv(filename, sep="\t").T
+    df_non_diploid = df_non_diploid.T.iloc[0:, 3:]  # .rename_axis("#ID").reset_index()
+    new_c = df_non_diploid.iloc[0:, 0:]
```
This is an interesting perspective. I don't see a big difference between using 'sum' or 'counts'. I think having the code as it is (check below) can make further possible modifications easier.

```
    df_non_diploid = df_non_diploid.T.iloc[0:, 3:]
    new_c = df_non_diploid.iloc[0:, 0:]
    nondiploid_counts = new_c[new_c < 0.05].count()
```

epiercehoffman replied:
I don't really care about sum vs. counts, but I think it's best to do this without making unnecessary copies of the dataframe. Why do you need new_c if it copies all the rows and columns?

shadizaheri replied:
I used `new_c`  before printing out/testing the code. We can easily delete `new_c` and modify as:

`    nondiploid_counts = df_non_diploid[df_non_diploid < 0.05].count()
`

epiercehoffman replied:
Sounds good, let's go with that

epiercehoffman commented on src/sv-pipeline/scripts/make_evidence_qc_table.py:
```diff
@@ -0,0 +1,264 @@
+
+
+def read_wgd_scores(filename: str) -> pd.DataFrame:
+    return pd.read_csv(filename, sep="\t")
```
Can you name the output column `wgd_score`? Currently it is named `score` in the original WGD file, but that is ambiguous in the unified output table

shadizaheri replied:
Renamed to `wgd_score`.

epiercehoffman commented on src/sv-pipeline/scripts/make_evidence_qc_table.py:
```diff
@@ -0,0 +1,264 @@
+    df_non_diploid = df_non_diploid.T.iloc[0:, 3:]
+    new_c = df_non_diploid.iloc[0:, 0:]
+    nondiploid_counts = new_c[new_c < 0.05].count()
+    nondiploid_counts_df = pd.DataFrame(nondiploid_counts, columns=["nondiploid_counts"]).rename_axis(
```
Let's name this column `nondiploid_bins` to indicate where it came from
```suggestion
    nondiploid_counts_df = pd.DataFrame(nondiploid_counts, columns=["nondiploid_bins"]).rename_axis(
```

epiercehoffman commented on src/sv-pipeline/scripts/make_evidence_qc_table.py:
```diff
@@ -0,0 +1,264 @@
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "estimated_copy_number_filename",
```
Dashes are best practice for command-line arguments, rather than underscores. When you access these options later as variables the dashes will be automatically converted to underscores. \

I think since there are so many arguments they should be named rather than positional. You can add a short argument name if you want. 

Some of the files are marked optional in the WDL task - make sure that is handled on the script level.
```suggestion
        "--estimated-copy-number-filename",
```

shadizaheri replied:
@epiercehoffman , this is a valid point, thanks for sharing. I also think that since these are required arguments, not optional, it might be better/easier if we go with underscore here. You can read more [here](https://stackoverflow.com/questions/12834785/having-options-in-argparse-with-a-dash)

epiercehoffman replied:
You can make them required but still use `--`. I think that for this quantity of arguments they shouldn't all be positional.

shadizaheri replied:
@epiercehoffman , I think there could be positional for the python script, at the same time I agree that it's better to not have them as positional! Thanks for the comment. I applied your comment in [this](https://github.com/broadinstitute/gatk-sv/commit/0086235f678fb29d32c88aa9fdd15ba7a76a71b8) commit.

shadizaheri replied:
The files have been set as optional.

VJalili commented on src/sv-pipeline/scripts/make_evidence_qc_table.py:
```diff
@@ -0,0 +1,264 @@
+        df[ID_COL] = df.apply(lambda _: "NA", axis=0)
+        outlier_sample = df.pivot_table(columns=[ID_COL], aggfunc="size").astype(int)
+    else:
+        df[ID_COL] = df["Outlier_Sample"].str.split("/", expand=True)[idx].str.split(".", expand=True)[0]
```
Agree. I am more inclined to use [`.stem`](https://stackoverflow.com/a/47496703/947889), which should handle cases where you have multiple `.` in the filename, but `.split` should also be fine.

shadizaheri commented on src/sv-pipeline/scripts/make_evidence_qc_table.py:
```diff
@@ -0,0 +1,264 @@
+    new_c = df_non_diploid.iloc[0:, 0:]
+    nondiploid_counts = new_c[new_c < 0.05].count()
+    nondiploid_counts_df = pd.DataFrame(nondiploid_counts, columns=["nondiploid_counts"]).rename_axis(
+        "#ID").reset_index()
```
I see! I updated the script with ID_COL and checked to be sure that it's consistent all over the script.

shadizaheri commented on src/sv-pipeline/scripts/make_evidence_qc_table.py:
```diff
@@ -0,0 +1,264 @@
+        filename_manta: Outliers determined in EvidenceQC for Manta.
+        filename_melt: Outliers determined in EvidenceQC for MELT.
+        filename_wham: Outliers determined in EvidenceQC for Wham.
+        outlier_type:
```
thanks for the comment! This got modified [here](https://github.com/broadinstitute/gatk-sv/blob/sz_evidence_qc_python_script/src/sv-pipeline/scripts/make_evidence_qc_table.py#:~:text=outlier_type%3A%20low%20or%20high)

shadizaheri reviewed (commented):
thanks for the comment! This got modified [here](https://github.com/broadinstitute/gatk-sv/blob/sz_evidence_qc_python_script/src/sv-pipeline/scripts/make_evidence_qc_table.py#:~:text=outlier_type%3A%20low%20or%20high)

epiercehoffman commented on src/sv-pipeline/scripts/make_evidence_qc_table.py:
```diff
@@ -0,0 +1,264 @@
+    df_manta_high_outlier = read_outlier(filename_high_manta, get_col_name("manta", "high"))
+    df_melt_high_outlier = read_outlier(filename_high_melt, get_col_name("melt", "high"))
+    df_wham_high_outlier = read_outlier(filename_high_wham, get_col_name("wham", "high"), 11)
+    df_total_high_outliers = read_all_outlier(filename_high_manta, filename_high_melt, filename_high_wham, "high")
```
It is less efficient to read the file again from disk. Using the outputs of the previous function would decrease the number of times that you read from disk as well as the number of dataframe operations. You directly call and use read_outlier within read_all_outlier, so it would be very simple to make those dataframes inputs to read_all_outlier instead of calling the function again.

shadizaheri replied:
We are in some sense using the output of other functions (`read_outlier`) inside the `read_all_outlier`. I don't think that making this change is necessary at this stage as I don't see much of a difference here.

epiercehoffman replied:
Let's touch base offline, I think it will be easier to explain!

shadizaheri replied:
Sure! Meantime I thank you for your precise comment and conversation!
I think, for now, this is not going to cause any problems for us as I have checked the Pandas DataFrame memory usage for a different number of rows, columns, and data types. Even for the number of rows (in our case number of samples) in the order of 5e6 and the number of columns in the order of 40, we are not going to pass a couple of megabytes. 

So, I won't expect this to cause a problem for us, at the same time I definitely consider your comment for the future.

epiercehoffman commented on src/sv-pipeline/scripts/make_evidence_qc_table.py:
```diff
@@ -0,0 +1,274 @@
+        help="Sets the filename containing Melt QC outlier low.", nargs=1)
+
+    parser.add_argument(
+        "-w", "--wham-qc-outlier-low-filename",
```
-w is used twice as a short argument name, here and for wham high outlier
```suggestion
        "-q", "--wham-qc-outlier-low-filename",
```

shadizaheri replied:
very nice catch! I fixed it [here](https://github.com/broadinstitute/gatk-sv/blob/sz_evidence_qc_python_script/src/sv-pipeline/scripts/make_evidence_qc_table.py#:~:text=%22%2Dr%22%2C%20%22%2D%2Dwham%2Dqc%2Doutlier%2Dlow%2Dfilename%22%2C).

shadizaheri reviewed (commented):
Thanks for your work on this!

shadizaheri commented:
@epiercehoffman and @VJalili thank you both for your very great comments and help on this PR! 
I believe I have addressed all the comments here.
You can test the script by running 
`python make_evidence_qc_table.py -y estimated_copy_numbers.txt -d bwr_medianCov.transposed.bed -g bwr_WGD_scores.txt -b binwise_CNV_qValues.bed -z bwr.Manta.QC.outlier.high -e bwr.Melt.QC.outlier.high -w bwr.Wham.QC.outlier.high -a bwr.Manta.QC.outlier.low -s bwr.Melt.QC.outlier.low -r bwr.Wham.QC.outlier.low -m melt_insert_size.tsv -o testing` command line. I am also attaching a zip file containing the input files as and a sample of the output file to this comment.
Please note that `melt_insert_size.tsv` is a tsv file that I exported from the data table to test the python script. Because the information for 'melt_insert_size.tsv' is already included in the data table as an output from the GatherSampleEvidence step, most likely we will not need this once we write the wdl as we can read this info from the data table in Terra.
[table_EvidenceQC_test.zip](https://github.com/broadinstitute/gatk-sv/files/10143461/table_EvidenceQC_test.zip)

epiercehoffman commented on src/sv-pipeline/scripts/make_evidence_qc_table.py:
```diff
@@ -0,0 +1,269 @@
+        help="Sets the filename containing Wham QC outlier low.")
+
+    parser.add_argument(
+        "-m", "--melt-insert-size-filename",
```
Not necessary to merge.
Why: I think we should update every reference to this file in this script to call it "mean" insert size instead of "Melt". We currently calculate the insert size during MELT, but as we work towards deprecating MELT, that will likely change, so we should be more flexible/accurate with the naming here.
Also: "melt" > "mean" elsewhere in script
```suggestion
        "-m", "--mean-insert-size-filename",
```

epiercehoffman commented on src/sv-pipeline/scripts/make_evidence_qc_table.py:
```diff
@@ -0,0 +1,269 @@
+        help="Sets the filename containing bin-wise CNV q-values.")
+
+    parser.add_argument(
+        "-z", "--manta-qc-outlier-high-filename",
```
Not necessary to merge: Scramble outlier files are missing here, so we should make sure to come back and update this script to handle them later. 
In the mean time, please remove the scramble inputs in the WDL task since the script does not use them.

epiercehoffman commented on src/sv-pipeline/scripts/make_evidence_qc_table.py:
```diff
@@ -0,0 +1,269 @@
+    df_melt_low_outlier = read_outlier(filename_low_melt, get_col_name("melt", "low"))
+    df_wham_low_outlier = read_outlier(filename_low_wham, get_col_name("wham", "low"))
+    df_total_low_outliers = read_all_outlier(df_manta_low_outlier, df_melt_low_outlier, df_wham_low_outlier, "low")
+    df_melt_insert_size = read_melt_insert_size(filename_melt_insert_size)
```
Not necessary to merge / question of preference. When I tested without providing one of the outlier files, I noticed that it produced a column of NA values. I think I might prefer to omit the column if the corresponding file is not provided, but would be open to other views.

epiercehoffman commented on src/sv-pipeline/scripts/make_evidence_qc_table.py:
```diff
@@ -0,0 +1,269 @@
+        outlier_sample = df.pivot_table(columns=[ID_COL], aggfunc="size").astype(int)
+    outlier_df = outlier_sample.reset_index()
+    outlier_df.columns = [ID_COL, outlier_col_label]
+    return outlier_df
```
Not necessary to merge / question of preference. I think my preference would be to fill cells with 0, rather than `np.nan`, if the sample does not appear as an outlier at all in a specified category. This would allow the outlier columns to be stored as ints and may make downstream computation a bit simpler because users won't have to account for NaN values. But, I recognize that it's easier to have NaN values due to the dataframe merge later on, and I'm open to other views.

epiercehoffman reviewed (approved):
Thanks @shadizaheri for all the updates you've made this week and to both @shadizaheri and @VJalili for all your work on this! I think this is in good shape and can be merged - I've left a few comments of varying degrees of importance but I'm ok with merging as-is and returning to those later (more details below). I've tested it locally and it appears to work as expected. I'll include the details of my testing below for transparency.

Those are the important bits, the rest is details.

I've made a few comments, mainly related to required inputs/outputs. I have marked them as either "Address soon" if I believe they should ideally be addressed before merging the script (but ok if soon after) or "Not necessary to merge" if I think it's fine to come back to them in the future (without urgency). 

I approved the PR as-is because I think that when run as part of EvidenceQC in the way we typically expect (ie. with all of the inputs provided, except optionally omitting MELT outlier files), the script will perform as desired. So, the "Address soon" reflects that the script does not fully handle required/optional outputs correctly, but given the desire to get this PR in quickly, I would approve a merge as-is as long as we come back and fix these items soon to provide more flexibility with inputs. (Some comments relate to the WDL task as well so it would be good to read them over for those details even if we merge as-is.)

Testing details:
- Ran with all input files and checked: number of rows in output, no empty sample IDs, all values in expected range for metric, outlier counts add up across rows, rounded chrX as expected based on chrX CN (spot-checked), column names are good
- Spot-checked a couple of columns against the input files to make sure nothing got scrambled
- Ran with a gzipped estimated CN file and checked that the output is identical, since the provided test files were not gzipped but in the WDL several of the input files will be gzipped
- Ran removing one command line argument each time and checked for expected behavior. Worked fine to omit MELT outlier files, but I've left a few comments about optional/required inputs in other cases
